### PR TITLE
Add stand-by cost

### LINKED ISSUE #657: Add stand-by cost for committable links and generators
## Describe the feature you'd like to see

The current implementation of costs is a pure proportional function with respect to the energy produced.
In mathematical term, the following conceptual function applies:
$cost = weighting * marginalcost * power$

However, most fuel-fired generators may be better described by a piece-wise linear function, or quadratic function, especially when generators are committable.
In this issue, we propose the following revised function:
$cost = weighting * (standbycost * status + marginalcost * power)$

The above function remains linear and it extends the previous formulation without being braking.
Soon a PR by Enrico may be opened to include this draft.

### PR DESCRIPTION
**Pull request by enricogiglio**
Closes #657 
 
## Changes proposed in this Pull Request

@davide-f and I have added stand-by costs for committable generators and links. We have also added a test for the feature we implemented and believe that the proposed changes are consistent with the rest of the codebase. We have not yet integrated the documentation to agree with you what needs to be included, after your review of the code.

## Checklist

- [ ] Code changes are sufficiently documented; i.e. new functions contain docstrings and further explanations may be given in `doc`.
- [x] Unit tests for new features were added (if applicable).
- [x] Newly introduced dependencies are added to `environment.yaml`, `environment_docs.yaml` and `setup.py` (if applicable).
- [x] A note for the release notes `doc/release_notes.rst` of the upcoming release is included.
- [x] I consent to the release of this PR's code under the MIT license.

### CODE REVIEW

**Review by FabianHofmann — CHANGES_REQUESTED** [review]
Thanks for the contribution @enricogiglio!

In order to make the change applicable to the `n.optimize()` function, we would need the change in `pypsa/optimization/optimize.py`, in particular in the `define_objective` function. Would you be okay with doing that as well? 

Here a non-tested hint on how to start:

```python 

    # stand-by cost
    comps = {"Generator", "Link"}
    for c in comps:
        com_i = get_committable_i(n, c)

        if com_i.empty:
            continue
        
        stand_by_cost = (
            get_as_dense(n, c, "stand_by_cost", sns)
            .loc[:, lambda ds: (ds != 0).any()]
            .mul(weighting, axis=0)
        )

        idx = stand_by_cost.columns
        status = n.model.variables[f"{c}-status"].loc[:, idx]
        m.objective = m.objective + (status * stand_by_cost).sum()
```

**Comment by enricogiglio** [comment]
Thank you, @FabianHofmann! We have also integrated the code in optimize.py. We believe the proposed changes are compatible with the rest of the codebase.

**Comment by davide-f** [comment]
@FabianHofmann Beyond the specific code changes, I was wondering if and what may be appropriate to modify in the documentation.
Moreover, I haven't seen the updated changes including the quadratic penalty.
I was wondering if these two changes may be appropriate to include, maybe by:
1. defining a symbol for the generator and links costs
2. defining this symbol as the sum of the stand-by cost, the marginal and quadratic terms

or if it is not in the focus of this PR to address this

File: pypsa/optimization/optimize.py [hunk 1]
```diff
@@ -139,6 +139,31 @@ def define_objective(n, sns):
...
+        idx = stand_by_cost.columns
+        com_and_stand_by_cost = com_i.intersection(idx)
+        status = n.model.variables[f"{c}-status"].loc[:, com_and_stand_by_cost]
+        stand_by_cost.columns.set_names(
+            stand_by_cost.columns.name + "-com", inplace=True
+        )
+
+        m.objective = (
+            m.objective + (status * stand_by_cost.loc[:, com_and_stand_by_cost]).sum()
+        )
```
Comment (FabianHofmann) [inline pypsa/optimization/optimize.py:165]:
```suggestion
        stand_by_cost = (
            get_as_dense(n, c, "stand_by_cost", sns, com_i)
            .loc[:, lambda ds: (ds != 0).any()]
            .mul(weighting, axis=0)
        )

        stand_by_cost.columns.name = f"{c}-com"
        status = n.model.variables[f"{c}-status"].loc[:, stand_by_cost.columns]
        m.objective = (
            m.objective + (status * stand_by_cost).sum()
        )
```
Could you try out whether this refined option works?

Comment (FabianHofmann) [reply pypsa/optimization/optimize.py:165]: and thanks btw, looks nearly perfect :)

**Comment by FabianHofmann** [comment]
> @FabianHofmann Beyond the specific code changes, I was wondering if and what may be appropriate to modify in the documentation. Moreover, I haven't seen the updated changes including the quadratic penalty. I was wondering if these two changes may be appropriate to include, maybe by:
> 
> 1. defining a symbol for the generator and links costs
> 2. defining this symbol as the sum of the stand-by cost, the marginal and quadratic terms
> 
> or if it is not in the focus of this PR to address this

@davide-f Yes you are right, the documentation should be tackled as well. It can be time-consuming though, that's why I wouldn't expect people to do it right away. But we should not forget about it. If you are interested in doing that, you know, always welcome...

File: pypsa/optimization/optimize.py (hunk 1 above)
Comment (davide-f) [reply pypsa/optimization/optimize.py:165]: @enricogiglio could you try to run it?

Comment (enricogiglio) [reply pypsa/optimization/optimize.py]: It seems to be working and thank you very much @FabianHofmann :)

**Comment by davide-f** [comment]
@FabianHofmann Regarding the documentation, I'm wondering what comprehensive changes are expected and if you may be interested in addressing that after the changes have been stabilized, or if we shall propose something.
I am unsure about the stability of the current version, the styling you may be interested in having for the documentation and whether that documentation may take some time to reach stability. Unfortunately we all know that documentation is much needed but can easily time consuming
What do you think?

@enricogiglio may you have some time to dedicate here or do you have other priorities?

**Comment by enricogiglio** [comment]
To be honest, I have a somewhat busy schedule at the moment, and if @FabianHofmann says that updating the documentation for this PR is not a priority, I would rather focus on other tasks. However, if there is a need to update the documentation in the coming months, I will be happy to contribute to the proposed changes! :)

**Comment by FabianHofmann** [comment]
No that's fine, thanks @enricogiglio

**Review by fneum — APPROVED** [review]
LGTM

**Review by FabianHofmann — APPROVED** [review]

**Pull request merged**